syntax = "proto3";

package whisper;

service Whisper {
  rpc Whispering (WhisperRequest) returns (WhisperReply) {}
}

message WhisperRequest {
  oneof data {
    string file_path = 1;
    bytes file_content = 2;
  }
  repeated string passwords = 3;
}

message WhisperReply {
  repeated Node tree = 1;
}

message Meta {
  map<string, string> map_string = 1;
  map<string, int64> map_number = 2;
  map<string, bool> map_bool = 3;
}

message Node {
  int64 id = 1;
  int64 parent_id = 2;
  repeated int64 children = 3;
  oneof content {
      File file = 4;
      Data data = 5;
  }
  Meta meta = 6;
}

message File {
    string path = 1;
    string name = 2;
    int64 size = 3;
    string mime_type = 4;
    string extension = 5;
    string md5 = 6;
    string sha256 = 7;
    optional bytes content = 8;
}

message Data {
    string type = 1;
    bytes content = 2;
}
